# API Scavenger Hunt! 👀

## Introduction 👋
In this lab, you will learn the fundamentals of working with APIs as a data analyst by making `requests` to an API and formatting the `response` data appropriately. Each task will offer a different challenge, such as: 

- working with different request parameters and query parameters, 
- using API keys and tokens, and 
- eventually signing up for and making requests from the Open Weather Map API, then exporting data from that request to JSON and importing the JSON into Tableau.

## Resources 📕
We recommend having the following Student Hub articles open while making your requests
- [RESTful APIs](https://devcodecamp.helpkit.so/definitions-examples-and-frequently-asked-questions/vUgiURdNAgva1AaaH99uBY/rest-apis/oN6ZdGaNop8f52gx43kTj6)
- [Importing & Exporting JSON with Pandas](https://devcodecamp.helpkit.so/definitions-examples-and-frequently-asked-questions/vUgiURdNAgva1AaaH99uBY/importing--exporting-data-with-pandas/3gXNpDMk4fkWP6zdepRCod#json-files)

---
## Task 0: Imports
---
- ✅ In the code block below, import the `requests`, `pandas`, and `json` packages
    - > 👉🏼 The imports should be written as three different lines. Don't forget the alias for Pandas!

In [62]:
import requests
import pandas as pd
import json

---
## Task 1: Current Forecast
---
The Open Weather Map API is a popular, open source API for weather forecasts that includes future weather, historical weather data, pollution and air quality forecasting, and more!

For the purpose of this lab, we will focus on two main features of Open Weather Map API:
- Current Weather Data
- 5 Day / 3 Hour Forecast Data

We'll use this API to get practice with basic API usage and documentation parsing.

### Subtask 1: Create Open Weather Map API Account & Obtain Key
- ✅ Navigate to the following website and locate the `Sign In` page: 👉🏼 [Open Weather Map API](https://openweathermap.org/) 👈🏻
    - Click the option to **Create a new account**
- ✅ After creating your account, check your email for a verification message from Open Weather Map API
- ✅ Locate the API Key (Select your username in the top right --> My API Keys) and set it equal to a variable called `key` in the code block below. (Make sure it is a string!)

In [63]:
key = "8c8b1f93a0df6648c10188d4ce21b492"

### Subtask 2: Calculate Latitude & Longitude of Your Location
- ✅ Navigate to the following website: 👉🏼 [GPS Coordinates](https://www.gps-coordinates.net/my-location) 👈🏻
- ✅ When prompted, allow the browser to access your current location
- ✅ Note the values for Latitude and Longitude, and set them equal to the variables called `lat` and `lon` in the code block below. (Make sure it is a number!)

In [64]:
lat = 35.282031
lon = -120.668031

### Subtask 3: Generate Request URL
TODO
- ✅ Navigate to the following website  👉🏼 [API Documentation](https://openweathermap.org/api) 👈🏻
- ✅ Locate the documentation for the **Current Weather Data** API
- ✅ Copy the URL for this API, and store it in the `url` variable below.  Use a **Python f string** to insert your latitude, longitude, and API key values into the URL

In [65]:
url = f"https://api.openweathermap.org/data/2.5/weather?lat={lat}&lon={lon}&appid={key}"

### Subtask 4: GET your current weather forecast.  
TODO
- ✅ Use the above URL to make a GET request, and save the result as a variable
- ✅ "Prettify" the response & print it out, then open the output in a text editor for easy reference.  
    - > Use that output to answer the questions in the Markdown block below

In [66]:
response = requests.get(url).json()
pretty_response = json.dumps(response, indent=4, skipkeys=True)
print(pretty_response)

{
    "coord": {
        "lon": -120.668,
        "lat": 35.282
    },
    "weather": [
        {
            "id": 800,
            "main": "Clear",
            "description": "clear sky",
            "icon": "01d"
        }
    ],
    "base": "stations",
    "main": {
        "temp": 285.65,
        "feels_like": 284.91,
        "temp_min": 282.85,
        "temp_max": 287.33,
        "pressure": 1015,
        "humidity": 75
    },
    "visibility": 10000,
    "wind": {
        "speed": 3.58,
        "deg": 61,
        "gust": 5.36
    },
    "clouds": {
        "all": 0
    },
    "dt": 1672251236,
    "sys": {
        "type": 2,
        "id": 2011788,
        "country": "US",
        "sunrise": 1672240224,
        "sunset": 1672275477
    },
    "timezone": -28800,
    "id": 5392323,
    "name": "San Luis Obispo",
    "cod": 200
}


QUESTION 1
- ❓ What is the **wind speed** for your current location?
- 🅰️:3.58

QUESTION 2
- ❓ How would you access the value for **wind speed** within this response?
- 🅰️: `wind_speed = response["wind"]["speed"]`

QUESTION 3
- ❓ What is the **weather description** for your current location?
- 🅰️:"clear sky"

QUESTION 4
- ❓ How would you access the value for **weather description** within this response?
- 🅰️: `weather_description = response["weather"]["description"]`


---
## Task 2: 5 Day / 3 Hour Forecast
---

### Subtask 1: Find the Forecast URL
Todo
- ✅ Navigate to the following website  👉🏼 [API Documentation](https://openweathermap.org/api) 👈🏻
- ✅ Locate the documentation for the **5 Day / 3 Hour Forecast** API
- ✅ Copy the URL for this API, and store it in the `forecast_url` variable below.  Use a **Python f string** to insert your latitude, longitude, and API key values into the URL
- ✅ Add the following additional parameter to the **end** of `forecast_url`, to convert temperature to Fahrenheit 
    - > **`&units=imperial`**

In [67]:
forecast_url = f"https://api.openweathermap.org/data/2.5/forecast?lat={lat}&lon={lon}&appid={key}&units=imperial"

### Subtask 2: GET your 5-day weather forecast. 
TODO
- ✅ Use the above URL to make a GET request, and save the result as a variable
- ✅ "Prettify" the response & print it out, then open the output in a text editor for easy reference.  

In [68]:
forecast = requests.get(forecast_url).json()
pretty_forcast = json.dumps(forecast, indent=4, skipkeys=True)
print(pretty_forcast)

{
    "cod": "200",
    "message": 0,
    "cnt": 40,
    "list": [
        {
            "dt": 1672261200,
            "main": {
                "temp": 56.95,
                "feels_like": 55.49,
                "temp_min": 56.95,
                "temp_max": 60.17,
                "pressure": 1014,
                "sea_level": 1014,
                "grnd_level": 1006,
                "humidity": 67,
                "temp_kf": -1.79
            },
            "weather": [
                {
                    "id": 801,
                    "main": "Clouds",
                    "description": "few clouds",
                    "icon": "02d"
                }
            ],
            "clouds": {
                "all": 17
            },
            "wind": {
                "speed": 5.59,
                "deg": 25,
                "gust": 8.43
            },
            "visibility": 10000,
            "pop": 0,
            "sys": {
                "pod": "d"
            },
            "

### Subtask 3: Prepare Data for Cleaning
TODO
- ✅ Examine the outputted JSON data, and locate the property that holds a **list of weather forecasts**
- ✅ Save this list of weather forecasts as the variable `forecast_list` below

In [69]:
forecast_list = forecast["list"]

### Subtask 4: Data Extraction
TODO
- ✅ Loop through each value within `forecast_list`, and perform the following:
    - ✅ Using square brackets, extract the following properties into their own variable:
        - Timestamp
        - Temperature
        - "Feels Like" Temperature
        - Wind Speed
        - Humidity
    - ✅ Create a **Dictionary** with key:value pairs for each of the above properties
    - ✅ Append that dictionary to the `data_list` list variable

In [70]:
data_list = []
for item in forecast_list:
    time_stamp = item["dt_txt"]
    temperature = item["main"]["temp"]
    feels_like = item["main"]["feels_like"]
    wind_speed = item["wind"]["speed"]
    humidity = item["main"]["humidity"]
    item_dictionary = {"time_stamp": time_stamp, "temperature": temperature, 
    "feels_like": feels_like, "wind_speed": wind_speed, "humidity": humidity}
    data_list.append(item_dictionary)


### Subtask 5: DataFrame Creation & Munging
TODO
- ✅ Using the Pandas DataFrame `.from_records()` method, create a new DataFrame from `data_list`, and save as a variable
- ✅ Use the appropriate Pandas method to convert the `timestamp` column to DateTime format

In [76]:
df = pd.DataFrame.from_records(data_list)
df["time_stamp"] = pd.to_datetime(df["time_stamp"])
df.head()

time_stamp  temperature  feels_like  wind_speed  humidity
0 2022-12-28 21:00:00        56.95       55.49        5.59        67
1 2022-12-29 00:00:00        56.39       54.82        6.42        66
2 2022-12-29 03:00:00        51.91       50.00        5.99        68
3 2022-12-29 06:00:00        50.36       48.25        3.67        67
4 2022-12-29 09:00:00        50.59       48.27        2.19        62

### Subtask 6: Export Data to JSON
TODO
- ✅ Use the appropriate Pandas function to export your DataFrame to **JSON format**
    - Name your file `forecastdata.json`
    - Use the orientation **`records`** when outputting
    - Include the argument **`date_format="iso"`** to ensure all DateTime values are exported correctly!

In [77]:
df.to_json("forecastdata.json", orient="records", date_format="iso")

### Subtask 7: Data Visualization with Tableau
TODO
- ✅ Within Tableau, import your `forecastdata.json` file
- ✅ Ensure that the `timestamp` column has been converted to Date & Time
- ✅ Create **two** visualizations exploring trends and patterns in your area's 5-day forecast
- ✅ Save these visualizations to Tableau Public, then download a copy of the workbook, and include it in your repository folder